bin/get-video-s3-acls.py

```diff
@@ -0,0 +1,418 @@
+            "FROM dictionary_gloss AS dg JOIN video_glossvideo AS vg ON vg.gloss_id = dg.id"
+            ") TO STDOUT WITH DELIMITER AS '|'",
+            f"{DATABASE_URL}",
+        ],
+        env=os.environ,
+        capture_output=True,
+        check=True,
+        text=True,
+    )
+
+    # Separate the NZSL db columns
+    # Write them to a dictionary, so we can do fast operations
+    for rawl in result.stdout.split("\n"):
+        rawl = rawl.strip()
+        if not rawl:
+            continue
+        [
+            gloss_id,
+            gloss_idgloss,
+            created_at,
+            gloss_public,
+            video_public,
+            video_id,
+            video_key,
+        ] = rawl.split("|")
+        this_nzsl_raw_keys_dict[video_key] = [
+            gloss_id,
+            gloss_idgloss,
+            created_at,
+            gloss_public.lower() == "t",
+            video_public.lower() == "t",
+            video_id,
+        ]
+
+    print(
+        f"{len(this_nzsl_raw_keys_dict)} rows retrieved",
+        file=sys.stderr,
+    )
+
+    return this_nzsl_raw_keys_dict
+
+
+# Get the keys present and absent across NZSL Signbank and S3, to dictionary
+# See DICTIONARY and CACHE FILE format
+def create_all_keys_dict(this_s3_bucket_raw_keys_list, this_nzsl_raw_keys_dict):
+    print(
+        "Getting keys present and absent across NZSL Signbank and S3 ...",
+        file=sys.stderr,
+    )
+    this_all_keys_dict = {}
+    with open(ALL_KEYS_CACHE_FILE, "w") as cache_file:
+
+        # Find S3 keys that are present in NZSL, or absent
+        for video_key in this_s3_bucket_raw_keys_list:
+            if video_key in this_nzsl_raw_keys_dict:
+                # NZSL PRESENT, S3 PRESENT
+                this_all_keys_dict[video_key] = [True, True] + this_nzsl_raw_keys_dict[
+                    video_key
+                ]
+            else:
+                # NZSL Absent, S3 PRESENT
+                this_all_keys_dict[video_key] = [False, True, "", "", "", "", "", ""]
+
+        # Find NZSL keys that are absent from S3 (present handled already above)
+        for video_key, item_list in this_nzsl_raw_keys_dict.items():
+            if video_key not in this_s3_bucket_raw_keys_list:
+                # NZSL PRESENT, S3 Absent
+                this_all_keys_dict[video_key] = [True, False] + item_list
+
+        # Write all keys back to a cache file
+        for video_key, item_list in this_all_keys_dict.items():
+            cache_file.write(
+                f"{video_key}{CSV_DELIMITER}{CSV_DELIMITER.join(map(str, item_list))}\n"
+            )
+
+    return this_all_keys_dict
+
+
+def build_csv_header():
+    return CSV_DELIMITER.join(
+        [
+            "Gloss ID",
+            "Gloss",
+            "Created at",
+            "Gloss public",
+            "Video public",
+            "Video ID",
+            "Video key",
+            "Expected Canned ACL",
+            "Actual Canned ACL",
+        ]
```

Comment:
issue: can we reorder these to:

- S3 key
- S3 Created At
- Expected Canned ACL
- Actual Canned ACL
- Gloss ID
- Video ID
- Gloss Public
- Video Public

My reasoning is:
  - it's very more common for a file to exist in S3 without a gloss than the other way around (and the latter is a more serious issue anyway)
  - the canned ACL columns + the S3 key are really the most important ones; all the others are for digging into specific items as needed
  - the id and public fields are equivalent and the latter we want to compare, so it's useful to have them paired together by that rather than paired by entity
  - I don't think the "Gloss" gives us any extra information? though not against you sticking it on the end if you think that'll be useful

Reply:
Absolutely.

Reply:
Yes I think the Gloss could be handy, would like to keep it if possible.